Added new config response and optimized regex

a-czyrny opened the pull request:
Added new config response and optimized regex. See https://github.com/nficano/pytube/issues/806 & https://github.com/nficano/pytube/issues/807

Commit:
Added new config response and optimized regex

Added new config response and optimized regex. See https://github.com/nficano/pytube/issues/806 & https://github.com/nficano/pytube/issues/807

a-czyrny commented:
Tested with 32 videos (2 from each year since 2005).

Commit:
CI fix

# noqa: E501

tfdahlin commented on pytube/extract.py:
```diff
@@ -263,8 +264,7 @@ def get_ytplayer_config(html: str) -> Any:
     setconfig_patterns = [
-        r"yt\.setConfig\((.*'PLAYER_CONFIG':\s*{.+?})\);",
-        r"yt\.setConfig\((.*\"PLAYER_CONFIG\":\s*{.+?})\);"
+        r"yt\.setConfig\((.*['\"]PLAYER_CONFIG['\"]:\s*{.+?})\);"
```
Good catch. At one point while I was experimenting with these regexes I had this, not sure what happened to that.

tfdahlin commented on pytube/extract.py:
```diff
@@ -248,7 +248,8 @@ def get_ytplayer_config(html: str) -> Any:
     logger.debug("finding initial function name")
     config_patterns = [
-        r";ytplayer\.config\s*=\s*({.*?});",
+        r";ytplayer\.config\s*=\s*({.*?(?<!gdpr)});",
```
What's the context for the gdpr bit? Do you happen to have an example video where we need to capture this?

a-czyrny replied:
Unfortunately it happens rarely and not reliably that the "gdpr" appears in the dictionary. It is then enclosed by two curly brackets with a following semicolon. This would lead to a premature hit and the dictionary would be invalid. Therefore this string is ignored.

tfdahlin replied:
I'll take your word for it then.

tfdahlin commented on pytube/__main__.py:
```diff
@@ -123,8 +123,11 @@ def descramble(self) -> None:
         if 'streamingData' not in self.player_config_args['player_response']:
             config_response = get_ytplayer_config(
```
lines 124-126 can be compressed to a single line, `config_response = get_ytplayer_config(self.watch_html)` now that the logic for setting `self.player_config_args` has been split off of it

a-czyrny replied:
Indeed.

Commit:
Update __main__.py

Refactoring.

tfdahlin commented:
I'll merge this a little later today, just want to run it through a few test videos myself before I do.

a-czyrny commented:
> I'll merge this a little later today, just want to run it through a few test videos myself before I do.

All right. Wouldn't do otherwise.

tfdahlin commented:
Funnily enough, I'm getting an issue when I try `YouTube('https://www.youtube.com/watch?v=9bZkp7q19f0').streams.first.download()` occasionally. Going to try and figure out what's happening.

tfdahlin commented:
Oh, wait, I may not have installed your branch correctly. Let me try to fix that first.

tfdahlin commented:
No, still running into the issue. Gonna see if I can figure out why.

a-czyrny commented:
The deviating dictionary requires a close examination. I still get in rare cases an exception because it cannot be parsed correctly.

tfdahlin commented:
Let's see if we can resolve the issue before merging this then, I'll be able to dedicate more time to it this afternoon

tfdahlin commented:
Sorry, gonna have to do this tomorrow, some stuff came up, and I got distracted by another PR i was working on

a-czyrny commented:
Here is an example for a watch page with the "{gdpr};" string inside the ytplayer.config dictionary:

https://gist.github.com/a-czyrny/3406242315c2503a5971c667bed8c816

Commit:
Improvments

Regex pattern improved and json.loads is only executed when necessary. Fixes the latest exceptions (finally) and processes the 32 test videos reliably multiple times.

tfdahlin committed:
Change to regex. This should help with bounding the search better.

tfdahlin commented:
I think I found a better way to bound the end of the string we're looking for, it works successfully against the videos I've been testing it against locally, and still passes all of the existing unit tests. Can you see if you have any issues with it @a-czyrny? I removed the extra gdpr bit in one of the regexes because I don't think it will be relevant any more with this new regex.

a-czyrny commented:
@tfdahlin Your solution does not have the flexibility of my solution. If the string does not end with " ytplayer" there is no match although the dictionary is present (see the new ytInitialPlayerResponse for example). In my solution it doesn't matter what comes after the semicolon.  If necessary the special feature of "gdpr" will be handled as well.

tfdahlin commented:
@a-czyrny i was running into issues with your regex, notably with the gangnam style video, where the json wouldn't decode properly. When I looked at the the raw HTML for pages that didn't parse correctly, I was consistently seeing the json dictionary end, and then another ytplayer element get set:
![image](https://user-images.githubusercontent.com/9667869/98956249-ccee5880-24cd-11eb-9965-3f74d86c0757.png)
This seemed to be the case with all of the videos that I was testing against, and this change also passed all of the unit tests successfully without needing to make changes. The string that we're looking for, where `ytplayer.config` gets set, is part of a larger <script> tag that also sets `ytplayer.web_player_context_config` and `ytplayer.load`.

a-czyrny commented:
The feedback on my adjustments is very positive (https://github.com/nficano/pytube/issues/815#issuecomment-726350383, https://github.com/nficano/pytube/issues/815#issuecomment-726432525) so I see no reason to change them. Otherwise I need the URL that leads to an error.

tfdahlin commented:
Both of those comments came *after* I made the changes (I pushed my change 3 days ago, those comments were yesterday) :p they're using the version with my regex in it

a-czyrny commented:
I missed that. So speaks for you. Nevertheless I would like to see an example where my regex pattern breaks. I still see the danger that your solution with the fixed extension/order in the string will become obsolete faster if only a small change is made (e.g. the order in the script, another following key etc.).

tfdahlin commented:
For me, it regularly broke on this video: https://www.youtube.com/watch?v=9bZkp7q19f0
It was a little inconsistent, but after a try or two i ran into the issue

tfdahlin commented:
```python
from pytube import YouTube, Playlist
from urllib import parse
import sys, traceback

test_ids = [
    'DkffTDkSADI',  # Indochine - Song for a Dream (Clip officiel)
    'XhFzKhTFY3I',  # Indochine - La vie est belle (Clip officiel)
    'fJ9rUzIMcZQ',  # Queen - Bohemian Rhapsody
    '9bZkp7q19f0',  # PSY - GANGNAM STYLE(강남스타일) M/V
    'UGN00k2bHgs',  # Hustle (P!nk)
    'Zi_XLOBDo_Y',  # Billie Jean
    'dQw4w9WgXcQ',  # Never Gonna Give You Up
]

bad_urls = []
exceptions = []
tracebacks = []

for id in test_ids:
    try:
        url = f'https://www.youtube.com/watch?v={id}'
        yt = YouTube(url)
        #yt.streams.first().download(skip_existing=False)
    except Exception as e:
        #with open(f'bad_html_{id}.html', 'wb') as f:
        #    f.write(yt.watch_html.encode('utf-8'))
        exceptions.append(e)
        tracebacks.append(traceback.format_exc())
        bad_urls.append(url)

for i in range(len(exceptions)):
    print(bad_urls[i])
    print(exceptions[i])
    print(tracebacks[i]) 
```
This is the code I was using for testing, I think I also had problems with the billie jean video

a-czyrny commented:
Thanks for the example, I will debug with it later.

a-czyrny commented:
I found a new regex pattern `>\s*?yt.setConfig\(({.+?})\);` that's leads to: https://gist.github.com/a-czyrny/7dc53b22189c7f958b50d462829f7d1c. For example the key length_seconds/lengthSeconds is changed to videoDurationSeconds. None of my patterns fail no matter how often I run it, only the new pattern (untreated) fails.

tfdahlin commented:
was this on an age-restricted video by chance?

a-czyrny commented:
Matches for "Indochine - La vie est belle (Clip officiel)".

tfdahlin commented:
Yeah, I suspected that may be an age-restricted video. It *may* be worth refactoring some of the code to extract different things based on whether the video is age restricted or not. It sounds like the embed html might differ enough to warrant that

a-czyrny commented:
I agree. There are at least 3 dictionaries of which one has different keys.

a-czyrny reviewed (commented):
Can we agree that your changes are not "necessary" and we leave it at my variant. The missing patterns have to be implemented separately as discussed and throw an exception for the time being.

tfdahlin commented:
Yeah, lets revert my changes, merge this, and work on a new PR for the different variants

tfdahlin committed:
Partial revert, leaving the \s* in place of \s+?

tfdahlin commented:
@a-czyrny I reverted my change, but the test suite I was using failed on three of the videos:
https://www.youtube.com/watch?v=DkffTDkSADI
https://www.youtube.com/watch?v=XhFzKhTFY3I
https://www.youtube.com/watch?v=Zi_XLOBDo_Y

I found out that there were a couple similar cases to what you found with the `{gdpr};` example, but with different values:
`{GDPR};`
`{ivc_transaction}`
Because these are different lengths, we can't simply change your look-behind to something like `(?<!(gdpr)|(GDPR)|(ivc_transaction))`, as look-behinds require fixed-width patterns.
However, I don't run into these problems at all with the addition of `;ytplayer` to the end of the string like I do with this reversion. I think that for the time being, my solution of adding `;ytplayer` to the end is more effective given the current state of youtube.

I tried looking into things a little bit more to see if we could potentially create a more robust regex for this, but then I realized that a simple regex can't solve this problem -- we'll need a context-aware parser for js. That's a very heavy lift though, and should be a stretch goal.

a-czyrny commented:
Then keep your last change. Thank you for the effort.

tfdahlin committed:
Added regex boundary back in. Need to investigate why some videos return no match.

tfdahlin committed:
Removed unused import.

tfdahlin committed:
Added better exception handling in js_url function.
